File: src/agent/outbox.ts
import AsyncStorage from '@react-native-async-storage/async-storage';
import { nanoid } from '@/agent/utils/nanoid';

const STORAGE_KEY = 'riflett_outbox_v1';

export type OutboxJobKind = 'polish';

export interface OutboxJob {
  id: string;
  kind: OutboxJobKind;
  payload: Record<string, unknown>;
  createdAt: number;
}

const loadJobs = async (): Promise<OutboxJob[]> => {
  try {
    const raw = await AsyncStorage.getItem(STORAGE_KEY);
    if (!raw) return [];
    const parsed = JSON.parse(raw) as OutboxJob[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
};

const saveJobs = (jobs: OutboxJob[]): Promise<void> =>
  AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(jobs));

export const Outbox = {
  async queue(job: Omit<OutboxJob, 'id' | 'createdAt'> & { id?: string }): Promise<OutboxJob> {
    const current = await loadJobs();
    const payload: OutboxJob = {
      id: job.id ?? nanoid(),
      kind: job.kind,
      payload: job.payload,
      createdAt: Date.now(),
    };
    current.push(payload);
    await saveJobs(current);
    return payload;
  },

  async list(): Promise<OutboxJob[]> {
    return loadJobs();
  },

  async clear(jobId: string): Promise<void> {
    const jobs = await loadJobs();
    const filtered = jobs.filter((job) => job.id !== jobId);
    await saveJobs(filtered);
  },
};

File: services/python-ai/app/agent/outbox.py
import asyncio
import time
from typing import Any, Dict, List, Literal, Optional

from .utils.nanoid import nanoid

OutboxJobKind = Literal["polish"]


class OutboxJob(Dict[str, Any]):
    id: str
    kind: OutboxJobKind
    payload: Dict[str, Any]
    createdAt: int


_storage: List[OutboxJob] = []
_queue_lock = asyncio.Lock()


async def _load_jobs() -> List[OutboxJob]:
    return list(_storage)


async def _save_jobs(jobs: List[OutboxJob]) -> None:
    _storage.clear()
    _storage.extend(jobs)


class Outbox:
    @staticmethod
    async def queue(job: Dict[str, Any]) -> OutboxJob:
        async with _queue_lock:
            current = await _load_jobs()
            payload: OutboxJob = {
                "id": job.get("id") or nanoid(),
                "kind": job["kind"],
                "payload": job["payload"],
                "createdAt": int(time.time() * 1000),
            }
            current.append(payload)
            await _save_jobs(current)
            return payload

    @staticmethod
    async def list() -> List[OutboxJob]:
        return await _load_jobs()

    @staticmethod
    async def clear(job_id: str) -> None:
        async with _queue_lock:
            jobs = await _load_jobs()
            filtered = [job for job in jobs if job.get("id") != job_id]
            await _save_jobs(filtered)

